File: csharp/AoC2021/Day25/Day25/SeaFloor.cs
namespace Day25;

public class SeaFloor
{
    private readonly List<SeaCucumber> _seaCucumbers;
    private readonly int _rows;
    private readonly int _cols;

    private SeaFloor Empty() => new(
        Enumerable.Repeat(SeaCucumber.Empty, _rows * _cols).ToList(),
        _rows,
        _cols
    );

    public bool TryNext(out SeaFloor next)
    {
        next = Empty();
        var anyoneMoved = false;

        for (var i = 0; i < _seaCucumbers.Count; i += 1)
        {
            switch (_seaCucumbers[i])
            {
                case SeaCucumber.East:
                {
                    var destination = GetRowOffset(i) + GetCol(i + 1);
                    if (_seaCucumbers[destination].IsEmpty())
                    {
                        next._seaCucumbers[destination] = SeaCucumber.East;
                        anyoneMoved = true;
                    }
                    else
                    {
                        next._seaCucumbers[i] = SeaCucumber.East;
                    }

                    break;
                }
                case SeaCucumber.South:
                {
                    var destination = (i + _cols) % (_rows * _cols);
                    var left = GetRowOffset(destination) + GetCol(destination - 1);
                    var right = GetRowOffset(destination) + GetCol(destination + 1);

                    var isFree = _seaCucumbers[destination].IsEmpty() && !_seaCucumbers[left].IsEast();
                    var willBeFree = _seaCucumbers[destination].IsEast() && _seaCucumbers[right].IsEmpty();

                    if (isFree || willBeFree)
                    {
                        next._seaCucumbers[destination] = SeaCucumber.South;
                        anyoneMoved = true;
                    }
                    else
                    {
                        next._seaCucumbers[i] = SeaCucumber.South;
                    }

                    break;
                }
            }
        }

        return anyoneMoved;
    }

    private int GetRowOffset(int i) => i - (i % _cols);

    private int GetCol(int i) => (i < 0 ? i + _rows * _cols : i) % _cols;

    private SeaFloor(List<SeaCucumber> seaCucumbers, int rows, int cols)
    {
        _seaCucumbers = seaCucumbers;
        _rows = rows;
        _cols = cols;
    }

    public static SeaFloor Parse(IEnumerable<string> input)
    {
        var cucumbers = new List<SeaCucumber>();

        var rows = 0;
        var cols = 0;
        foreach (var line in input)
        {
            rows += 1;
            cols = 0;
            foreach (var ch in line)
            {
                cols += 1;
                cucumbers.Add(ch switch
                {
                    '>' => SeaCucumber.East,
                    'v' => SeaCucumber.South,
                    _ => SeaCucumber.Empty,
                });
            }
        }

        return new SeaFloor(cucumbers, rows, cols);
    }
}

File: rust/2021/day25/src/sea_floor.rs
use std::io;

use grid::{grid, Grid};
use rdcl_aoc_helpers::error::ParseError;
use rdcl_aoc_helpers::grid::iterators::WithGridIterator;

use crate::sea_cucumber::SeaCucumber;

pub trait SeaFloor {
    fn next(&self) -> Option<Self>
    where
        Self: Sized;

    fn parse_input<I>(lines: I) -> Result<Self, ParseError>
    where
        Self: Sized,
        I: Iterator<Item = io::Result<String>>;
}

impl SeaFloor for Grid<SeaCucumber> {
    /// Determines the next generation of the sea floor.
    fn next(&self) -> Option<Self>
    where
        Self: Sized,
    {
        let rows = self.rows();
        let cols = self.cols();
        let mut next_value = Grid::new(rows, cols);
        let mut anyone_moved = false;

        for (y, x) in self.iter_row_col() {
            match self[y][x] {
                SeaCucumber::East => {
                    let nx = (x + 1) % cols;
                    if self[y][nx].is_empty() {
                        next_value[y][nx] = SeaCucumber::East;
                        anyone_moved = true;
                    } else {
                        next_value[y][x] = SeaCucumber::East;
                    }
                }
                SeaCucumber::South => {
                    let px = (x + cols - 1) % cols;
                    let nx = (x + 1) % cols;
                    let ny = (y + 1) % rows;

                    let target_is_free = self[ny][x].is_empty() && !self[ny][px].is_east();
                    let target_will_be_free = self[ny][x].is_east() && self[ny][nx].is_empty();

                    if target_is_free || target_will_be_free {
                        next_value[ny][x] = SeaCucumber::South;
                        anyone_moved = true;
                    } else {
                        next_value[y][x] = SeaCucumber::South;
                    }
                }
                SeaCucumber::Empty => {}
            }
        }

        if anyone_moved {
            Some(next_value)
        } else {
            None
        }
    }

    /// Parses the sea floor from the puzzle input.
    fn parse_input<I>(lines: I) -> Result<Self, ParseError>
    where
        Self: Sized,
        I: Iterator<Item = io::Result<String>>,
    {
        let mut grid = grid![];
        for line in lines {
            grid.push_row(
                line?
                    .chars()
                    .map(|ch| match ch {
                        '>' => SeaCucumber::East,
                        'v' => SeaCucumber::South,
                        _ => SeaCucumber::Empty,
                    })
                    .collect(),
            );
        }
        Ok(grid)
    }
}
